geo/src/algorithm/euclidean_distance.rs

```diff
@@ -480,47 +397,229 @@ where
     }
 }
 
-// ┌──────────────────────────────────┐
-// │ Implementations for MultiPolygon │
-// └──────────────────────────────────┘
+// ┌─────────────────────────────┐
+// │ Shared implementations      │
+// └─────────────────────────────┘
 
-impl<T> EuclideanDistance<T, Point<T>> for MultiPolygon<T>
-where
-    T: GeoFloat,
-{
-    /// Minimum distance from a MultiPolygon to a Point
-    fn euclidean_distance(&self, point: &Point<T>) -> T {
-        point.euclidean_distance(self)
+/// Implements EuclideanDistance between Geometry<T> and specific geometry types
+/// by converting the specific types to Geometry<T> and using the Geometry<T> implementation.
+/// Note that currently this macro cannot be used for GeometryCollection because Geometry::from(GeometryCollection) is currently disabled.
+macro_rules! impl_euclidean_distance_to_geometry {
+  ([$($for:ty),*]) => {
+      $(
+          impl<T> EuclideanDistance<T, Geometry<T>> for $for
+          where
+              T: GeoFloat + FloatConst + RTreeNum + FloatConst,
+          {
+              fn euclidean_distance(&self, geom: &Geometry<T>) -> T {
+                  let self_as_geom = Geometry::from(self.clone());
```

Comment:
This usage of clone is unfortunate, since it potentially doubles our memory usage. Did you explore alternatives that don't require a `clone` here? Maybe it'd be a ton of code?

Reply:
Thanks for the review! I've updated this to avoid the use of clone

Reply:
I'm quite personally interested in this :) Could the use of cloning here be avoided by creating the `impl`s manually?

Reply:
@JosiahParry the most recent commit removes this use of clone :)